src/libos/src/time/timer_file.rs

```diff
@@ -0,0 +1,394 @@
+            // already trigger the initial timer
+            _ => {
+                // Starts from the initial expire time,
+                // Calculate the elapsed total time and expected next expire time.
+                // The diff of above two is the time to next expire.
+                let elapsed = cur_time.checked_sub(init_exp).unwrap();
+                let next_expire = interval.checked_mul((exp_cnt + 1) as u32).unwrap();
+
+                ret_time.it_value = match next_expire.checked_sub(elapsed) {
+                    Some(duration) => duration,
+                    _ => {
+                        warn!("Error calculating timerfd time until next expire");
+                        Duration::new(0, 0)
+                    }
+                };
+            }
+        };
+
+        Ok(ret_time)
+    }
+
+    fn start_timer_task(&self, it_value: Duration) -> Result<()> {
+        let mut timeout = it_value;
+        let inner_c = self.inner.clone();
+        let mut inner = inner_c.lock().unwrap();
+        let interval = inner.interval;
+
+        // Check if the async loop block needs break (one_shot = true)
+        let one_shot = match inner.status {
+            TimerFdStatus::LOOP => false,
+            _ => true,
+        };
+
+        // Start background poll task to monitor timerfd events
+        let join_handle = SpawnOptions::new({
+            let pollee = self.pollee.clone();
+            let inner_c = self.inner.clone();
+
+            async move {
+                loop {
+                    let waiter = Waiter::new();
+                    // The initial timeout is the it_value.
+                    let res = waiter.wait_timeout(Some(&mut timeout)).await;
+                    match res {
+                        Err(e) => {
+                            if e.errno() == EINTR {
+                                // Error or Stopped by tirq
+                                break;
+                            } else if one_shot == true {
+                                // One-shot timer, no need loop
+                                trace!("timerfd one-shot triggerred");
+                                pollee.add_events(Events::IN);
+                                let mut inner = inner_c.lock().unwrap();
+                                inner.exp_cnt += 1;
+                                inner.status = TimerFdStatus::EXPIRE;
+                                inner.task_handle = None;
+                                break;
+                            } else {
+                                trace!("timerfd timer expired");
+                                // If it is not one-shot, the following timeout changes to interval.
+                                timeout = interval;
+                                pollee.add_events(Events::IN);
+                                let mut inner = inner_c.lock().unwrap();
+                                inner.exp_cnt += 1;
```

Comment:
We could combine the common code:
      pollee.add_events(Events::IN);
      let mut inner = inner_c.lock().unwrap();
      inner.exp_cnt += 1;